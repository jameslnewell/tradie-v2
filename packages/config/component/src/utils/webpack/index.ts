import * as fs from 'fs';
import * as path from 'path';
import * as Webpack from 'webpack';
import * as HtmlWebpackPlugin from 'html-webpack-plugin';

const DEBUG = Boolean(process.env.DEBUG);

function getTemplate(root: string) {
  const defaultTemplatePath = path.join(__dirname, 'template.html');
  const userTemplatePath = path.join(root, 'src/index.html');
  if (fs.existsSync(userTemplatePath)) {
    return userTemplatePath;
  } else {
    return defaultTemplatePath;
  }
}

export function getWebpackConfig(root: string): Webpack.Configuration {

  const typescriptConfigFile = require.resolve('../typescript/tsconfig.website.json');
  const tsxEntryFile = path.join(root, 'website/src/index.tsx');
  const tsEntryFile = path.join(root, 'website/src/index.ts');

  return ({
    mode: 'development',

    devtool: 'source-map',

    context: root,

    entry: [
      fs.existsSync(tsxEntryFile) ? tsxEntryFile: tsEntryFile
    ],

    output: {
      filename: 'index.js'
    },

    resolve: {
      extensions: ['.ts', '.tsx', '.js'],

      // use "source" over the other built files to avoid having to run build after changing
      mainFields: ['source', 'browser', 'module', 'main'],

      // give users a nice name for accessing the project-info loader
      alias: {
        '@tradie/project-info': require.resolve('./project-info-placeholder.txt')
      }

    },

    module: {
      rules: [
        {
          test: /project-info-placeholder.txt/,
          exclude: /node_modules/,
          loader: require.resolve('./project-info-loader'),
        },
        {
          test: /\.tsx?$/,
          exclude: /node_modules/,
          loader: require.resolve('ts-loader'),
          options: {
            configFile: typescriptConfigFile,
            logLevel: DEBUG ? 'debug' : 'silent',
            silent: DEBUG ? false : true,
            logInfoToStdOut: false,
            transpileOnly: false, // TODO: write own plugin to do typechecking
            errorFormatter: (error: {line: string; character: string; content: string;}) => {
              return `${error.line}:${error.character} ${error.content}`;
            }

          }
        }
      ]
    },

    plugins: [
      new HtmlWebpackPlugin({ // TODO: use package name for page title
        template: getTemplate(root)
      })
    ]

  });

}
