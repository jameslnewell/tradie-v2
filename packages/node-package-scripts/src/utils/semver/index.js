import {Range, lt} from 'semver';

export function min(arg) {
  let version = null;

  //FIXME: this is a VERY buggy method (rough guess) that doesn't take into account the operator
  new Range(arg).set.forEach(comparator => {
    comparator.forEach(range => {
      if (range.operator === '=' || range.operator === '>' || range.operator === '>=') {
        if (version === null) {
          version = range.semver;
          return;
        }

        if (lt(range.semver, version)) {
          version = range.semver;
          return;
        }
      }
    });
  });

  return version && version.raw;
}
