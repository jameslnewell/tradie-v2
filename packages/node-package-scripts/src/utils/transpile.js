export default function(transpiler) {
  return (file, report) =>
    transpiler.transpile(file).catch(error => report.error(file, error, 1));
}
