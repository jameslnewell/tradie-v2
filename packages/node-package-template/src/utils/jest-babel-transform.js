import {createBabelTransform} from '@tradie/jest-utils';
import {getBabelConfig} from './babel';

/* eslint-disable import/no-commonjs */
module.exports = createBabelTransform(getBabelConfig({root: process.cwd()}));
/* eslint-enable import/no-commonjs */
