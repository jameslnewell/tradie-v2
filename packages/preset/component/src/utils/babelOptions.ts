
export function cjs() {
  return {
    babelrc: false,
    sourceMaps: true,
    presets: [
      [
        require.resolve('@babel/preset-env'),
        {
          modules: 'commonjs',
          useBuiltIns: 'usage'
        }
      ],
      require.resolve('@babel/preset-typescript')
    ],
    plugins: [
      require.resolve('@babel/plugin-proposal-object-rest-spread'),
      require.resolve('@babel/plugin-proposal-class-properties')
    ]
  };
}

export function esm() {
  return {
    babelrc: false,
    sourceMaps: true,
    presets: [
      [
        require.resolve('@babel/preset-env'),
        {
          modules: false,
          useBuiltIns: 'usage'
        }
      ],
      require.resolve('@babel/preset-typescript')
    ],
    plugins: [
      require.resolve('@babel/plugin-proposal-object-rest-spread'),
      require.resolve('@babel/plugin-proposal-class-properties')
    ]
  };
}
