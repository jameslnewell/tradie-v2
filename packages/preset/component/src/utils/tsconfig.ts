import * as tsconfig from '../../tsconfig.json';

export function sources() {
  return {
    ...tsconfig,
    compilerOptions: {
      ...tsconfig.compilerOptions,
      emitDeclarationOnly: true
    }
  };
}

export function tests() {
  return {
    ...tsconfig,
    compilerOptions: {
      ...tsconfig.compilerOptions,
      noEmit: true
    }
  };
}
