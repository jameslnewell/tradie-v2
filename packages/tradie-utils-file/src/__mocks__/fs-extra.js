const fs = jest.genMockFromModule('fs-extra');

fs.__files = {};

fs.ensureDir = jest.fn().mockReturnValue(Promise.resolve());

fs.unlink = jest.fn(file => {
  delete fs.__files[file];
  return Promise.resolve();
});

fs.readFile = jest.fn(file => {
  if (typeof fs.__files[file] === undefined) {
    return Promise.reject(new Error(`File "${file}" not found.`));
  }
  return Promise.resolve(new Buffer(fs.__files[file]));
});

fs.writeFile = jest.fn((file, contents) => {
  fs.__files[file] = Buffer.isBuffer(contents) ? contents.toString() : contents;
  return Promise.resolve();
});

export default fs;
