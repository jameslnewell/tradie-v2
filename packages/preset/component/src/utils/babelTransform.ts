// @ts-ignore
import {createTransformer} from 'babel-jest';
import {cjs} from './babelOptions';

module.exports = createTransformer(cjs());
