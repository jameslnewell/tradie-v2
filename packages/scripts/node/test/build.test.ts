import * as path from 'path';
import * as fs from 'fs-extra';
import {createFixture, exec, read, cleanFixtures} from './utils';

describe('tradie build', () => {
  jasmine.DEFAULT_TIMEOUT_INTERVAL = 20000;

  afterAll(() => cleanFixtures())

  it('should exit with a code of 0 when there are no errors', async () => {
    const fixtureDir = await createFixture('sum');
    const {code, stdout, stderr} = await exec('build', {cwd: fixtureDir});
    expect(code).toEqual(0);
  });

  it('should transpile source when there are no errors', async () => {
    const fixtureDir = await createFixture('sum');
    const {code} = await exec('build', {cwd: fixtureDir});
    expect(code).toEqual(0);
    expect(await read(path.join(fixtureDir, 'lib/index.js'))).toMatchSnapshot();
  });

  it('should export types when there are no errors', async () => {
    const fixtureDir = await createFixture('sum');
    const {code} = await exec('build', {cwd: fixtureDir});
    expect(code).toEqual(0);
    expect(await read(path.join(fixtureDir, 'lib/index.d.ts'))).toMatchSnapshot();
    expect(await read(path.join(fixtureDir, 'lib/index.js.flow'))).toMatchSnapshot();
  });

  it('should copy other files', async () => {
    const fixtureDir = await createFixture('sum');
    const {code} = await exec('build', {cwd: fixtureDir});
    expect(code).toEqual(0);
    expect(await read(path.join(fixtureDir, 'lib/data.json'))).toMatchSnapshot();
  });

  it('should exit with a code of 1 when there are lint errors', async () => {
    const fixtureDir = await createFixture('lint-error');
    const {code} = await exec('build', {cwd: fixtureDir});
    expect(code).toEqual(1);
  });

  it('should exit with a code of 1 when there are syntax errors', async () => {
    const fixtureDir = await createFixture('syntax-error');
    const {code} = await exec('build', {cwd: fixtureDir});
    expect(code).toEqual(1);
  });

  it('should exit with a code of 1 when there are type errors', async () => {
    const fixtureDir = await createFixture('type-error');
    const {code} = await exec('build', {cwd: fixtureDir});
    expect(code).toEqual(1);
  });
});
