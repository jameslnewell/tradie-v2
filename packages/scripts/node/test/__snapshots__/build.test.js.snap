// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tradie build should copy other files when there are no errors 1`] = `
"{
  \\"sum\\": 99
}
"
`;

exports[`tradie build should export types when there are no errors 1`] = `
"// @flow

export default function(...args: number[]) {
  return args.reduce((total, arg) => total + arg, 0);
}
"
`;

exports[`tradie build should transpile source when there are no errors 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});

exports.default = function () {
  for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
    args[_key] = arguments[_key];
  }

  return args.reduce(function (total, arg) {
    return total + arg;
  }, 0);
};"
`;
